solvers/SimpleSOL/Diagnostics/momentum_conservation.hpp

```diff
@@ -0,0 +1,307 @@
+    const auto pl_stride =
+        particle_group->mpi_rank_dat->get_particle_loop_cell_stride();
+    const auto pl_npart_cell =
+        particle_group->mpi_rank_dat->get_particle_loop_npart_cell();
+
+    sycl_target->queue
+        .submit([&](sycl::handler &cgh) {
+          cgh.parallel_for<>(
+              sycl::range<1>(pl_iter_range), [=](sycl::id<1> idx) {
+                NESO_PARTICLES_KERNEL_START
+                const INT cellx = NESO_PARTICLES_KERNEL_CELL;
+                const INT layerx = NESO_PARTICLES_KERNEL_LAYER;
+
+                const double contrib = k_ion_p1[cellx][0][layerx];
+
+                sycl::atomic_ref<double, sycl::memory_order::relaxed,
+                                 sycl::memory_scope::device>
+                    energy_atomic(k_particle_momentum_1[0]);
+                energy_atomic.fetch_add(contrib);
+
+                NESO_PARTICLES_KERNEL_END
+              });
+        })
+        .wait_and_throw();
+
+    this->dh_particle_total_momentum_1.device_to_host();
+    const double tmp_particle_momentum_1 = this->dh_particle_total_momentum_1.h_buffer.ptr[0];
+    double total_particle_momentum_1;
+    MPICHK(MPI_Allreduce(&tmp_particle_momentum_1, &total_particle_momentum_1, 1, MPI_DOUBLE,
+                         MPI_SUM, sycl_target->comm_pair.comm_parent));
+
+    return total_particle_momentum_1;
+  }
+  
+  
+   inline double compute_total_transferred_momentum_0() {
+    const double particle_momentum_transferred_0 =
+        this->particle_sys->total_particle_momentum_transferred[0];
+    double global_particle_momentum_transferred_0;
+    MPICHK(MPI_Allreduce(&particle_momentum_transferred_0, &global_particle_momentum_transferred_0, 1,
+                         MPI_DOUBLE, MPI_SUM,
+                         sycl_target->comm_pair.comm_parent));
+
+
+    return global_particle_momentum_transferred_0;
+  } 
+  
+   inline double compute_total_transferred_momentum_1() {
+    const double particle_momentum_transferred_1 =
+        this->particle_sys->total_particle_momentum_transferred[1];
+    double global_particle_momentum_transferred_1;
+    MPICHK(MPI_Allreduce(&particle_momentum_transferred_1, &global_particle_momentum_transferred_1, 1,
+                         MPI_DOUBLE, MPI_SUM,
+                         sycl_target->comm_pair.comm_parent));
+
+
+    return global_particle_momentum_transferred_1;
+  }   
+  
+  
+  
+   inline double compute_total_added_momentum_0() {
+    const double particle_momentum_added_0 =
+        this->particle_sys->total_particle_momentum_added[0];
+    double global_particle_momentum_added_0;
+    MPICHK(MPI_Allreduce(&particle_momentum_added_0, &global_particle_momentum_added_0, 1,
+                         MPI_DOUBLE, MPI_SUM,
+                         sycl_target->comm_pair.comm_parent));
+
+
+    return global_particle_momentum_added_0;
+  } 
+  
+  
+  
+  
+
+
+  inline double compute_fluid_0_momentum() {
```

Comment:
I think you should generalise this class for n dimensions (where the number of dimensions are passed in the constructor). Otherwise we will end up with 2,3 copies of the same function with the only difference being an index.